Memory_analyzer library converted to complete void-safety.

jocelyn opened the pull request.

jocelyn committed:
Memory_analyzer library converted to complete void-safety.

manuseiffel commented on Src/library/memory_analyzer/ma_memory_change_mediator.e:
```diff
@@ -82,14 +88,21 @@ feature -- Command
 			-- Adjust the split area (which ia at the top area)'s position.
 		do
-			main_window.split_incre.set_split_position ((main_window.split_incre.width / 2).ceiling)
+			if attached main_window as l_main_window then
```
Do we need this, the attribute is attached?

jocelyn replied:
no

manuseiffel commented on Src/library/memory_analyzer/ma_draw_history.e:
```diff
@@ -15,6 +15,12 @@ inherit
 		export
 			{NONE} all
 		end
+
```
Should we pass this as argument to the creation procedure instead? Like it is done elsewere?

jocelyn replied:
I didn't want to change too much the current design.
I just changed it for one interface that was using a lot the main_window.

Now I can do that for all callers of main_window, but this means change the related attribute to be pseudo once per object (I mean using internal_foo: detachable like foo ..., and foo: FOO do ....-- use internal_foo end)

manuseiffel commented on Src/library/memory_analyzer/singletons/ma_shared_main_window.e:
```diff
@@ -0,0 +1,68 @@
+
+feature -- Access
+
+	main_window: detachable MA_WINDOW
```
If we can avoid it, I would prefer. If we have to keep it, let's make it attached by default. Meaning that the MA_WINDOW should be created right away in the once.

jocelyn replied:
Note this new class is just the previous design (in MA_SINGLETON_FACTORY) splitted into two classes.
I added this class, since it was not really the same nature of other once from MA_SINGLETON_FACTORY, and I guess the MA_SINGLETON_FACTORY.main_window was more a workaround due to void-safety and the difficulty to use Current to initialize attached attributes.

manuseiffel commented on Src/library/memory_analyzer/singletons/ma_singleton_factory.e:
```diff
@@ -69,20 +69,6 @@ feature  -- Singletons
 			system_util_not_void: Result /= Void
 		end
 
-	main_window: MA_WINDOW
```
Was there a reason to extract this code in MA_SHARED_MAIN_WINDOW? Couldn't just update and use this class instead?

jocelyn replied:
See my previous comment, in MA_SHARED_MAIN_WINDOW.
And in addition, it was to avoid conflict with inherited `main_window`.

manuseiffel commented on Src/library/memory_analyzer/widgets/ma_constants_imp.e:
```diff
@@ -2,8 +2,8 @@ note
+	revision: "$Revision: 91071 $"
 
 class
 	MA_CONSTANTS_IMP
```
This is generated by EiffelBuild. Did you regenerate it or manually updated it?

jocelyn replied:
Just updated manually, in fact I forgot this was generated by EiffelBuild.
Maybe this is old generated code since .. this is clearly not complete void-safe.

jocelyn committed:
Merge branch 'master' of https://github.com/EiffelSoftware/EiffelStudio into voidsafe_memory_analyzer

jocelyn committed:
Updated conversion to complete void-safety.

jocelyn committed:
remove date and revision change to minimize diff.

jocelyn committed:
Cosmetic and revert some changes on MA_CONSTANTS_IMP.

jocelyn commented:
The code was updated after comments.

manuseiffel commented on Src/library/memory_analyzer/ma_draw_history.e:
```diff
@@ -45,7 +50,9 @@ feature -- Command
 		do
 			internal_pixmap.clear
-			internal_pixmap.set_size (main_window.gc_graphs.width, main_window.gc_graphs.height)
+			if attached main_window as l_main_window then
```
Do we need the object test here since the attribute is attached?

jocelyn replied:
no I forgot to remove.

manuseiffel commented on Src/library/memory_analyzer/ma_gc_info_mediator.e:
```diff
@@ -19,134 +20,144 @@ create
-			main_window.generation_object_limit_other.set_text (memory.generation_object_limit.out)
-			main_window.max_memory_other.set_text (memory.max_mem.out)
-		end	
+			if attached main_window as l_main_window then
```
Is object test necessary since attribute is attached?

jocelyn replied:
no, fixed

manuseiffel commented on Src/library/memory_analyzer/ma_memory_change_mediator.e:
```diff
@@ -87,9 +93,15 @@ feature -- Command
 		do
-			main_window.split_incre_horizontal.set_split_position (main_window.split_incre_horizontal.minimum_split_position
-				+ main_window.memory_spot_1.row_count * main_window.memory_spot_1.row_height + main_window.memory_spot_1.header.height)
+			l_main_window := main_window
```
No need for the local here.

jocelyn replied:
this was just to avoid attribute access again and again.

manuseiffel commented on Src/library/memory_analyzer/ma_memory_change_mediator.e:
```diff
@@ -280,36 +293,37 @@ feature {NONE} -- Implemention
 			l_grid_data_increased := grid_data_increased
-			check attached l_grid_data_increased end -- Implied by precondition
-			l_sorter.sort (l_grid_data_increased)
+			if l_grid_data_increased = Void then
```
Use /= Void first and put check in else part.

jocelyn replied:
ok

manuseiffel commented on Src/library/memory_analyzer/ma_memory_change_mediator.e:
```diff
@@ -367,13 +378,10 @@ feature {NONE} -- Implemention
 			check False end -- Anchor type only
-			check attached l_result end -- Satisfy void-safe compiler
-			Result := l_result
+			;(create {EXCEPTIONS}).die (-1)
```
Use check False then end as discussed.

jocelyn replied:
done

manuseiffel commented on Src/library/memory_analyzer/ma_memory_state.e:
```diff
@@ -163,26 +164,23 @@ feature {NONE} -- Implementation
-			check l_result /= void end -- Satisfy void-safe compiler
-			Result := l_result
+				-- Used as anchor type!
+			;(create {EXCEPTIONS}).die (-1)
```
ditto

jocelyn replied:
done

manuseiffel commented on Src/library/memory_analyzer/ma_object_snapshot_mediator.e:
```diff
@@ -125,8 +125,11 @@ feature -- Command
 				if finding_route_to_once then
-					check attached l_map end -- Implied by `finding_route_to_once'
-					add_to_reference_table (l_new_table.key_for_iteration, l_map)
+					if l_map = Void then
```
Put check in else clause instead.

jocelyn replied:
done

manuseiffel commented on Src/library/memory_analyzer/ma_object_snapshot_mediator.e:
```diff
@@ -232,27 +235,30 @@ feature {NONE} -- Implementation
-					-- Perform update action.
-				a_update_action.call (Void)
+			l_grid_data := grid_data
+			if l_grid_data = Void then
```
put check in else clause instead.

jocelyn replied:
done

manuseiffel commented on Src/library/memory_analyzer/ma_object_snapshot_mediator.e:
```diff
@@ -272,10 +278,13 @@ feature {NONE} -- Implementation
 		do
 			from
 				l_data := grid_data
+				if l_data = Void then
```
Looks to me that the whole loop should be surrounded by `if l_data /= Void then`. We don't even need the else clause here since the precondition will caught it.

manuseiffel commented on Src/library/memory_analyzer/ma_object_snapshot_mediator.e:
```diff
@@ -676,12 +689,9 @@ feature {NONE} -- Fields
-			check attached l_result end -- Satisfy void-safe compiler
-			Result := l_result
+				-- Should not be used, except for anchor type
+			;(create {EXCEPTIONS}).die (-1)
```
Use check False then end

jocelyn replied:
done

manuseiffel commented on Src/library/memory_analyzer/memory_analyzer.ecf:
```diff
@@ -24,4 +21,9 @@
 		<cluster name="stone" location="stone\" recursive="true"/>
 		<cluster name="widgets" location="widgets\" recursive="true"/>
 	</target>
+	<target name="mem" extends="memory_analyzer">
```
Do we need this target?

jocelyn replied:
no I removed it, this was to compile in non void-safe mode while converting.
This won't be committed.

manuseiffel commented on Src/library/memory_analyzer/object_graph/ma_figure_factory.e:
```diff
@@ -60,17 +62,19 @@ feature -- Basic operations
 --				end
 --			end
+			to_implement (generator + ".model_from_xml: Not yet implemented!")
+			;(create {EXCEPTIONS}).die (-1)
```
Use `check False then end`

jocelyn replied:
In this case, I am not sure .. this is not implemented, so a "die" sounds good to me.

manuseiffel replied:
If no void-safety is required, non-implemented features are either implemented with require/check/ensure False.

jocelyn replied:
I am not sure to see what you mean

manuseiffel commented on Src/library/memory_analyzer/object_graph/ma_object_graph_mediator.e:
```diff
@@ -172,26 +173,32 @@ feature -- Command
 			fig := world.figure_from_model (l_last_drawn_node)
-			check attached fig end -- Implied by `l_last_drawn_node' has been just added
-			fig.set_point_position (ax, ay)
+			if fig = Void then
```
Put check in else clause

jocelyn replied:
done

manuseiffel commented on Src/library/memory_analyzer/object_graph/ma_reference_link.e:
```diff
@@ -105,43 +105,50 @@ feature {EG_FIGURE, EG_FIGURE_WORLD} -- Update
-				if is_label_shown then
-					name_label.set_point_position (line.x, line.y)
-				end
+			if l_model = Void then
```
Put check in else clause

jocelyn replied:
done.
Note that I initially did that, since it is sometime a pain with big if then else end code .. to see the associated condition for the "else" code.
with ` = Void ` we see it right away, and the else is not far from the `if`.

however I understand that here ... the `check ... end ` should never be reached so it sounds fair to put it as "else".

manuseiffel commented on Src/library/memory_analyzer/singletons/ma_icons_singleton.e:
```diff
@@ -38,7 +38,10 @@ feature {NONE} -- Implementation
 		do
 			l_result := internal_pixmap_path
-			check attached l_result end -- FIXME: Implied by ...?
+			if l_result = Void then
```
Could we remove this and instead make the `set_pixmap_directory` a creation procedure that sets `internal_pixmap_path`?

jocelyn replied:
Not that simple for now.
Maybe we can try after the conversion to complete void-safety.

jocelyn replied:
Finally I made pixmap_path detachable.
And updated the unique caller.

manuseiffel commented on Src/library/memory_analyzer/widgets/ma_window.e:
```diff
@@ -38,17 +39,14 @@ feature {NONE} -- Initialization
 
 			create_all_widgets
 			create_object_routes_panel_objects
+			create timer
```
Looks like those changes are incorrect. You should simply redefine `create_interface_objects` and add creations there, no?

jocelyn replied:
ok

manuseiffel commented on Src/library/memory_analyzer/ma_references_table.e:
```diff
@@ -67,14 +67,16 @@ feature -- Element change
 			l_hash: detachable like references_by_referee
 		do
-			if not relations.has_key (a_referee) then
+			if
```
Replace by `attached relations.item (a_referee)`

manuseiffel commented on Src/library/memory_analyzer/ma_references_table.e:
```diff
@@ -85,12 +87,11 @@ feature -- Removal
-			if relations.has_key (a_referee) then
-				l_hash := relations.found_item
-				check l_hash /= Void end -- Implied by `has_key'
+			if
```
Same here `attached relations.item (a_referree)`

jocelyn replied:
done

jocelyn committed:
Updated after comments
Mainly removed useless object tests.
Use `check False then end` and `callable: False` for function used for anchor type purpose.

jocelyn committed:
removed useless inheritance

jocelyn committed:
Better code for not implemented function.

jocelyn commented:
Close since committed.
